src/Perl6/Metamodel/C3MRO.nqp

```diff
@@ -1,10 +1,15 @@
 role Perl6::Metamodel::C3MRO {
     # Storage of the MRO.
     has %!mro;
+    has $!mro_lock;
+
+    method setup_mro_engine() {
+        $!mro_lock := NQPLock.new();
+    }
```

Comment:
If this returns `self`, then all all the `.new` methods would just have to add a call to `setup_mro_engine`, and not need a local `$obj`.

Reply:
True and I was considering it. But there is better solution, though it involves NQP patching. NQP build plan already supports `BUILD` method. But since `BUILD` replaces the entire default build plan, adding support for `TWEAK` would make it a good solution. In this case a few locations in the Metamodel could be replaced with a series of `self.setup_<something>` calls.

Aside of this, the body of `new` is already a bit hard to read at the first glance. Adding a method call wouldn't be beneficial.

Reply:
Why just not `has $!mro_lock := NQPLock.new();` since it looks like you call this method in all the `.new()`s anyway?

Reply:
@MasterDuke17 Maintainability.

Reply:
> @MasterDuke17 Maintainability.

How? Doing it in one place instead of several seems more maintainable to me. Unless you think a bunch of new Perl6::Metamodel::C3MRO subclasses aren't going to want to protect the hash?

Reply:
Oh, I misread your initial comment. It's NQP, it doesn't support attribute initializers. Wish it would, but if we go this way we're going to end up with self-bootstrapping Rakudo. Oh, wait, why not? ;)